### ISSUE TITLE
Create GDAL 3.3/3.4+ proj 8.2 builds against icu68

### ISSUE THREAD

**Issue by akrherz** [conda-forge/gdal-feedstock#585]
I think there is a fair amount of dependency log jam within conda-forge due to the stuck icu69 migration for qt,  conda-forge/qt-feedstock#207.  It appears significant work is happening within the qt repo and I am unsure of how long it may take.

My 2 cents is that we could help this situation out greatly by creating new GDAL 3.3 and 3.4 builds with proj 8.2 and against icu 68.  I am willing to do that work, but am unsure of the "blessed" approach.  My ignorant idea is to:

1. PR a reversal of the icu69 migration with build number bump.
2. Merge that
3. Migrate icu69 again with build bump and merge that

**Comment by akrherz** [comment]
As I debug other repos with dep troubles, I find this puzzling combo, which may be helped with this change...

```
$ mamba create -n dev python=3.10 proj=8.1.1 gdal=3.3 geos=3.10.0 icu=68
...
  - package gdal-3.3.3-py37h7b489ff_1 requires python_abi 3.7.* *_cp37m, but none of the providers can be installed
```

**Comment by akrherz** [comment]
Well, maybe not as we would also need to temp revert proj 8.2.0 migration :)

**Comment by akrherz** [comment]
The above maybe just a quirk of the mamba solver, indeed

libgdal-3.3.3-hf5a45a9_1.tar.bz2 -> `icu >=68.2,<69.0a0` + `proj >=8.1.1,<8.1.2.0a0`

So this change would not help those still stuck on proj 8.1.1, but proj 8.2.0 migrations would work.

**Comment by gillins** [comment]
How many projects are held up by this? If it's only `qgis` I'd say let's wait as the current `qgis` version being built is so old I suspect no one is using it...

**Comment by akrherz** [comment]
> How many projects are held up by this?

Fair question and likely not enough to fuss about, but am trying to get migrations finished out as I can and am seeing the icu69 issue about everywhere I poke around these days.

**Comment by xylar** [comment]
@akrherz, I think this will be a huge headache and likely isn't worth the trouble. 

Regarding your comment https://github.com/conda-forge/gdal-feedstock/issues/585#issuecomment-984089768, I think that strange failure message just means that there isn't a solution.  As you know, the `mamba` output isn't always very helpful or intuitive on why.

**Comment by akrherz** [comment]
> I think this will be a huge headache

Thanks @xylar for the response.  Could you kindly and tersely describe this headache?  From my perspective, we merge my two currently green PRs, the bot submits to more PRs that will be green to merge, and we will be done, hopefully :)

**Comment by xylar** [comment]
The bot won't re-migrate `gdal` once it considers it to be done, so we will have to migrate back to icu69 manually as well.  But what I really think will be a headache is that the problem isn't just with `gdal`, it's with all feedstocks that migrated to `icu69` that are dependencies of `qgis`.

It's also a messy business to migrate and migrate back that wastes a lot of CI time, particularly for `gdal`, which takes so long.  I personally just don't think it's worth it.

**Comment by xylar** [comment]
But I guess you've already done more than half of the work related to `gdal` on this so I'm willing to stand aside and let you proceed.

**Comment by akrherz** [comment]
Thanks @xylar , I didn't realize that about the bot.  So how about this.

1) I'll do the manual icu69 migration afterwards
2) I'll agree to not dig this hole any deeper if this icu68 idea does not pan out as I expect it to.

Deal?! :)

**Comment by xylar** [comment]
Yep, sounds okay.

**Comment by akrherz** [comment]
Thanks.  

- [x] verified `v3.3.x` branch build number 11 was successful (gdal=3.3.3)
- [x] verified `master` branch build number 9 was successful (gdal=3.4.0)

:eyes: repo metadata update and will log here if my suspicion was successful or not with unjamming downstream repos.

**Comment by akrherz** [comment]
As an example of what I am attempting to fix here, the [conda-forge status page](https://conda-forge.org/status/#other_migrations) reports not-solvable for the proj820 migration on `aarch64` like so:

```
r-sf (1) Error: not solvable (bot CI job): master:
 ['linux_64_r_base4.1: Encountered problems while solving:
 - package r-base-4.1.0-h9e01966_0 requires icu >=68.1,<69.0a0, but none of the providers can be installed
```

So r-base is still migrating icu69 conda-forge/r-base-feedstock#194 

and r-sf has a libgdal + proj dependency... and as I debugged this, the bot worked :)  conda-forge/r-sf-feedstock#63

**Comment by xylar** [comment]
Sorry for standing in the way earlier.  I didn't realize how much was jammed up by this (and that not much else was standing in the way).

**Comment by akrherz** [comment]
No need to apologize @xylar , while this change was helpful, I was too optimistic about how much it would fix.  I am still not entirely certain if conda-forge/qt-feedstock#207 is what is jamming everything up yet or not.  Mamba takes some interesting dependence paths that I am not fully certain of yet.  More debugging to come!

**Comment by akrherz** [comment]
lolz, I know I said I would go no further down this rabbit hole, but comparing for funzies:

```
$ mamba create -n dev python=3.9  geos=3.10.1 libgdal=3.3 qt
- package libgdal-3.3.3-h18e3bf0_2 requires geos >=3.10.0,<3.10.1.0a0, but none of the providers can be installed

$ mamba create -n dev python=3.9  geos=3.10.0 libgdal=3.3 qt
works

$ mamba create -n dev python=3.9  geos=3.10.1 libgdal=3.4 qt
- package qt-5.12.9-h9d6b050_2 requires libpq >=12.3,<13.0a0,

$ mamba create -n dev python=3.9  geos=3.10.0 libgdal=3.4 qt
- package libgdal-3.4.0-h913cfbe_4 requires icu >=69.1,<70.0a0
```

**Comment by akrherz** [comment]
OK, lets do a brain dump here as I have no idea what is going on with the `qt` repo.  Lets try to isolate what exactly is jamming things up, maybe `icu68/icu69` isn't it :)  The latest build of `qt` is 5.12.9, so we shall hard code that to mamba to attempt to better isolate things.

```
$ mamba create -n dev python=3.9 qt=5.12.9
Works.

$ mamba create -n dev python=3.9 qt=5.12.9 libgdal=3.4.0
- package qt-5.12.9-h1f2b2cb_0 requires icu >=67.1,<68.0a0
```

So, is there a qt build with icu=68, yes, qt-5.12.9-h9d6b050_2 (openssl 1) and qt-5.12.9-hda022c4_4 (openssl 1), OK, so openssl 3 has no path here.  Lets pick on that build 4 of QT now.

```
$ mamba create -n dev python=3.9 qt=5.12.9=hda022c4_4 libgdal=3.4.0
- package qt-5.12.9-hda022c4_4 requires libpq >=13.1,<14.0a0

$ mamba create -n dev python=3.9 libpq=13 libgdal=3.4.0
works.
```

So libpq has requirements on `krb5`, `libgcc-ng`, `libzlib` , `openssl`, and `readline`.   Picking on build `libpq-13.5-hd57d9b9_0` we iterate.

```
$ mamba create -n dev python=3.9 libpq=13.5=hd57d9b9_0 libgdal=3.4.0 "krb5 >=1.19.2,<1.20.0a0" "libgcc-ng >=9.4.0" "libzlib >=1.2.11,<1.3.0a0" "openssl >=1.1.1l,<1.1.2a" "readline >=8.1,<9.0a0"
works.
```

What?  Lets go back to the qt requirements again and iterated one by one until we find this combo:

```
$ mamba create -n dev python=3.9 libgdal=3.4.0 "libpq >=13.1,<14.0a0" "icu >=68.1,<69.0a0"
- package libgdal-3.4.0-hf890fe0_9 requires libpq >=14.1,<15.0a0
```

Augh, icu is back again.  So is there a icu68 + libpq13 available?

```
$ mamba create -n dev python=3.9  "libpq >=13.1,<14.0a0" "icu >=68.1,<69.0a0"
works.
```

Grrrr, I added a gdal3.4 + icu68 build (libgdal-3.4.0-hf890fe0_9 openssl1), why no likey :(  Well, that one had a `icu >=68.2,<69.0a0` requirement, so 68.1 won't fulfill that :(

```
$ mamba create -n dev python=3.9 libgdal=3.4.0=hf890fe0_9 "libpq >=13.1,<14.0a0" "icu >=68.2,<69.0a0"
- package libgdal-3.4.0-hf890fe0_9 requires openssl >=1.1.1l,<1.1.2a
```

```
$ mamba create -n dev python=3.9 "openssl >=1.1.1l,<1.1.2a" "libpq >=13.1,<14.0a0" "icu >=68.2,<69.0a0"
works
```

So something in ` libgdal=3.4.0=hf890fe0_9 is trouble?  And now I see the problem, gdal is compiled against "libpq >=14.1,<15.0a0"

```
$ mamba create -n dev python=3.9  qt=5.12.9  "libpq >=14.1,<15.0a0"
- package qt-5.12.9-h1f2b2cb_0 requires libpq >=12.3,<13.0a0
```

But, older qt works (5.12.5  hd8c4c69_1)

```
$ mamba create -n dev python=3.9  qt  "libpq >=14.1,<15.0a0"
works
```

That version of qt has no libpq requirement, sigh.  So now I wonder if it would be helpful for gdal to have a new icu68+libpq13 build :/

```
$ mamba create -n dev python=3.9  gdal=3.4 libpq=13 icu=68
- package gdal-3.4.0-py310h222ae0d_9 requires python_abi 3.10.* *_cp310
```

Anybody with thoughts for the good of my sanity? :)

**Comment by xylar** [comment]
@akrherz, I have a crude tool I use to debug these kinds of things that you might find handy as well:
https://gist.github.com/xylar/a506a16a474788b27e17f350eef547f8
The idea is that you put in the list of packages you're trying to debug, with the packages that are most important to you at the top.  It does a binary search to find the longest list of packages, starting from the beginning, that is solvable.  I typically put in all the runtime dependencies of a given version of a package I'm trying to debug (including the pins from `.ci_support`) and then sometimes need to replace a package with *its* dependencies.  Let me know if that helps you and also if you want someone else to help figure out the problem.

**Comment by akrherz** [comment]
Thanks @xylar , you are the best~  If you want to take a swing at this, but all means :)  The deal above is that I would not dig the rabbit hole any deeper and do another backdated rebuild, but it somebody else suggests it as a good idea, I would do the work :)

**Comment by akrherz** [comment]
And as an aside as to why I am motivated by this pain, I can't move forward with GDAL 3.4, because my users need GUI matplotlib and thus we are jammed up with `qt`.

**Comment by xylar** [comment]
I see, it's really helpful to understand the end goal.  So you want `gdal=3.4` and `qt` in the same environment?

**Comment by xylar** [comment]
I don't have time to help out right now but maybe over the weekend.  I have a bizarre fascination with puzzles like this one...

**Comment by akrherz** [comment]
@xylar , basically yes and such a combination will help out the various conda-forge repos that are jammed up for the same reason.  So I am attempting to kill two birds with one stone or at least use one as the excuse for resolving the other :/

**Comment by xylar** [comment]
@akrherz, my investigation suggests that the problem might be related to https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/1388.  The pin for `expat` got relaxed from `2.2.` to just `2` but I don't think `qt` got rebuilt with the more relaxed pin after that.  Do you want to try rebuilding `qt` with just a rerender to get the latest pins but without doing any other migrations?  That should pick up this relaxation.

**Comment by xylar** [comment]
In case you want to reproduce what I'm seeing: https://gist.github.com/xylar/e02438b5a19e8aa73439ef2bc42b9aea

**Comment by akrherz** [comment]
Thanks @xylar , but the `qt-feedstock` repo won't build at all.  see fun I having with conda-forge/qt-feedstock#210

**Comment by xylar** [comment]
Ah, well, that really is a can of worms.  It doesn't look like it's been getting much love, maybe for good reason.  Best of luck!

**Comment by akrherz** [comment]
I wasn't even able to get it `qt-feedstock` to build with only a build number bump and no re-render.  There's lots of love being given to the repo though...

**Comment by xylar** [comment]
I see.  I wasn't sure if it was getting attention since the last successful merge to `qt-feedstock` was 10 months ago: https://github.com/conda-forge/qt-feedstock/pull/183.  It sounds like it's getting attention, successful or otherwise.

**Comment by xylar** [comment]
@akrherz, I looked into the incompatibilities between the current build of `qt` and `gdal` a bit further.  If you want a build of `gdal=3.4.0` that works with `qt`, you would need:
```
expat >=2.2.10,<3.0.0a0
icu >=68.1,<69.0a0
libpq >=13.1,<14.0a0
```
My suggestion would be to make a build of `gdal` with those settings and then switch back to the defaults, similar to what you did for icu on its own.

**Comment by akrherz** [comment]
Thanks again @xylar , my reading of the expat pining is that all the qt=5.12.9 builds have: `"expat >=2.2.9,<3.0.0a0",` or `expat >=2.2.10,<3.0.0a0`.  How old of a qt did you get with your test that worked above?

**Comment by xylar** [comment]
@akrherz, you're right.  I was seeing issues with `expat` because I was explicitly pinning `expat=2.2`, which was what was in the `.ci_support` file.  But that seems to get rendered differently than I was expecting, as I commented in #590.

**Comment by akrherz** [comment]
Just to cross-reference a note that `libpq` never made it to conda-forge/conda-forge-pinning-feedstock#133

**Comment by akrherz** [comment]
We did it @xylar!!!

```
$ mamba create -n dev gdal=3.4.0 qt
...
gdal 3.4.0  py310h4b048a7_11
....
```

Will give the other arch repos some more time to update and then see about unjamming some migration PRs I have found :)

**Comment by xylar** [comment]
Okay, that's great!  Too bad about the weirdness in #591 but that seems to be unrelated to our shenanigans...

**Comment by akrherz** [comment]
closing as no appetite to repeat this with GDAL 3.4.1
